src/h3/mod.rs

```diff
@@ -642,19 +649,39 @@ impl Connection {
     /// a newly allocated stream.
     ///
     /// On success the newly allocated stream ID is returned.
+    ///
+    /// The [`StreamBlocked`] error is returned when the underlying QUIC stream
+    /// doesn't have enough capacity for the operation to complete. When this
+    /// happens the application should retry the operation once the stream is
+    /// reported as writable again.
+    ///
+    /// [`StreamBlocked`]: enum.Error.html#variant.StreamBlocked
```

Comment:
When a request attempt gets StreamBlocked, retrying the the request will allocate a new stream ID. This wastes a stream but also, depending on the conditions that trigger the blockage, a client might burn through a load of streams that have been allocated in memory but never reused.

Reply:
Oh, good point, though AFAICT this is already the case (e.g. if the server doesn't give enough stream ID credits).

I added an additional commit to only increase the stream IDs after the relevant operation complete (this also fixes the overflow checks, though I doubt anyone would ever be able to hit the limit anyway).